File: Java/SpaceVIL/spacevil/src/main/java/com/spvessel/spacevil/WindowLayout.java
package com.spvessel.spacevil;

import static org.lwjgl.system.MemoryUtil.NULL;

import java.awt.Color;
import java.awt.image.BufferedImage;
import java.util.UUID;
import java.util.concurrent.locks.Lock;
import java.util.concurrent.locks.ReentrantLock;

import com.spvessel.spacevil.Core.InterfaceOpenGLLayer;

final class WindowLayout {

    void restoreView() {
        if (_engine != null) {
            _engine.restoreView();
        }
    }

    void restoreCommonGLSettings() {
        if (_engine != null) {
            _engine.restoreCommonGLSettings();
        }
    }

    void setGLLayerViewport(InterfaceOpenGLLayer layer) {
        if (_engine != null) {
            _engine.setGLLayerViewport(layer);
        }
    }

    Lock engineLocker = new ReentrantLock();
    private Lock wndLock = new ReentrantLock();

    private UUID _id;

    UUID getId() {
        return _id;
    }

    void setId(UUID value) {
        _id = value;
    }

    private CoreWindow _coreWindow;

    CoreWindow getCoreWindow() {
        return _coreWindow;
    }

    void setCoreWindow() {
        _id = _coreWindow.getWindowGuid();

        wndLock.lock();
        try {
            WindowsBox.initWindow(_coreWindow);
            setFocusedItem(_window);
        } catch (Exception ex) {
            ex.printStackTrace();
            return;
        } finally {
            wndLock.unlock();
        }
    }

    private DrawEngine _engine;

    private Thread _threadActionManager;
    private ActionManager _actionManager;

    WindowLayout(CoreWindow window) {
        _coreWindow = window;
        _actionManager = new ActionManager(_coreWindow);
        _engine = new DrawEngine(_coreWindow);
        _coreWindow.eventClose.add(this::close);
    }

    private WContainer _window;

    WContainer getContainer() {
        return _window;
    }

    void setWindow(WContainer window) {
        _window = window;
    }

    final void show() {
        if (!WindowManager.isRunning())
            WindowManager.startWith(_coreWindow);
        else
            WindowManager.addWindow(_coreWindow);
    }

    void close() {
        WindowManager.closeWindow(_coreWindow);
    }

    void setFocusable(boolean value) {
        _engine.glwHandler.focusable = value;
    }

    void minimize() {
        _engine.minimizeRequest = true;
    }

    void maximize() {
        _engine.maximizeRequest = true;
    }

    void toggleFullScreen() {
        _engine.fullScreenRequest = true;
    }

    void updatePosition() {
        _engine.updatePositionRequest = true;
    }

    void updateSize() {
        _engine.updateSizeRequest = true;
    }

    void isFixed(Boolean flag) {
    }

    void setEventTask(EventTask task) {
        _actionManager.addTask(task);
    }

    void executePollActions() {
        _actionManager.execute.set();
    }

    Prototype getFocusedItem() {
        return _engine.getFocusedItem();
    }

    void setFocusedItem(Prototype item) {
        _engine.setFocusedItem(item);
    }

    void resetItems() {
        _engine.resetItems();
    }

    void resetFocus() {
        _engine.resetFocus();
    }

    void setIcon(BufferedImage icon_big, BufferedImage icon_small) {
        _engine.setBigIcon(icon_big);
        _engine.setSmallIcon(icon_small);
    }

    void setHidden(Boolean value) {
        _engine.glwHandler.setHidden(value);
    }

    long getGLWID() {
        return _engine.glwHandler.getWindowId();
    }

    boolean isGLWIDValid() {
        if (getGLWID() == NULL)
            return false;
        return true;
    }

    boolean initEngine() {
        if (_coreWindow.isHidden)
            setHidden(false);
        _coreWindow.isClosed = false;

        _engine.glwHandler.transparent = _coreWindow.isTransparent;
        _engine.glwHandler.maximized = _coreWindow.isMaximized;
        _engine.glwHandler.visible = !_coreWindow.isHidden;
        _engine.glwHandler.resizeble = _coreWindow.isResizable;
        _engine.glwHandler.borderHidden = _coreWindow.isBorderHidden;
        _engine.glwHandler.appearInCenter = _coreWindow.isCentered;
        _engine.glwHandler.focusable = _coreWindow.isFocusable;
        _engine.glwHandler.alwaysOnTop = _coreWindow.isAlwaysOnTop;

        _engine.glwHandler.getPointer().setX(_coreWindow.getX());
        _engine.glwHandler.getPointer().setY(_coreWindow.getY());

        _threadActionManager = new Thread(() -> {
            _actionManager.startManager();
        });
        _threadActionManager.setDaemon(true);
        _threadActionManager.start();

        createWindowsPair();

        return _engine.init();
    }

    void createWindowsPair() {
        WindowsBox.createWindowsPair(_coreWindow);
        if (_coreWindow.isDialog) {
            WindowsBox.getWindowPair(_coreWindow).setFocusable(false);
        }
    }

    void dispose() {
        CoreWindow currentPair = getPairForCurrentWindow();
        destroyWindowsPair();
        if (_threadActionManager != null && _threadActionManager.isAlive()) {
            _actionManager.stopManager();
            _actionManager.execute.set();
        }
        _engine.freeOnClose();
        if (currentPair != null && !currentPair.isClosed)
            currentPair.setWindowFocused();
    }

    void destroyWindowsPair() {
        if (_coreWindow.isDialog) {
            CoreWindow pair = WindowsBox.getWindowPair(_coreWindow);
            if (pair != null)
                pair.setFocusable(true);
            WindowsBox.removeWindow(_coreWindow);
        }
        WindowsBox.removeFromWindowDispatcher(_coreWindow);
    }

    CoreWindow getPairForCurrentWindow() {
        return WindowsBox.getWindowPair(_coreWindow);
    }

    void updateScene() {
        _engine.drawScene();
    }

    void render() {
        _engine.render();
    }

    void setFocus() {
        setFocusable(true);
        _engine.setWindowFocused();
    }

    private Color _shadeColor = new Color(0, 0, 0, 200);

    void setShadeColor(Color color) {
        wndLock.lock();
        try {
            _shadeColor = new Color(color.getRed(), color.getGreen(), color.getBlue(), color.getAlpha());
        } finally {
            wndLock.unlock();
        }
    }

    void setShadeColor(int r, int g, int b) {
        setShadeColor(GraphicsMathService.colorTransform(r, g, b));
    }

    void setShadeColor(int r, int g, int b, int a) {
        setShadeColor(GraphicsMathService.colorTransform(r, g, b, a));
    }

    void setShadeColor(float r, float g, float b) {
        setShadeColor(GraphicsMathService.colorTransform(r, g, b));
    }

    void setShadeColor(float r, float g, float b, float a) {
        setShadeColor(GraphicsMathService.colorTransform(r, g, b, a));
    }

    Color getShadeColor() {
        wndLock.lock();
        try {
            return _shadeColor;
        } finally {
            wndLock.unlock();
        }
    }

    <T> void freeVRAMResource(T resource) {
        _engine.freeVRAMResource(resource);
    }

    ToolTipItem getToolTip() {
        return _engine.getToolTip();
    }
}

File: SpaceVIL/Windows/WindowLayout.cs
﻿using System;
using System.Threading;
using System.Threading.Tasks;
using System.Drawing;
using SpaceVIL.Core;

namespace SpaceVIL
{
    internal sealed class WindowLayout
    {
        internal void RestoreView()
        {
            if (_engine != null)
            {
                _engine.RestoreView();
            }
        }

        internal void RestoreCommonGLSettings()
        {
            if (_engine != null)
            {
                _engine.RestoreCommonGLSettings();
            }
        }

        internal void SetGLLayerViewport(IOpenGLLayer layer)
        {
            if (_engine != null)
            {
                _engine.SetGLLayerViewport(layer);
            }
        }

        internal Object EngineLocker = new Object();
        private Object wndLock = new Object();

        private Guid _id;

        Guid getId()
        {
            return _id;
        }

        void setId(Guid value)
        {
            _id = value;
        }

        private CoreWindow _coreWindow;

        internal CoreWindow GetCoreWindow()
        {
            return _coreWindow;
        }

        internal void SetCoreWindow()
        {
            _id = _coreWindow.GetWindowGuid();
            Monitor.Enter(wndLock);
            try
            {
                WindowsBox.InitWindow(_coreWindow);
                SetFocusedItem(_window);
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.ToString());
                return;
            }
            finally
            {
                Monitor.Exit(wndLock);
            }
        }

        private DrawEngine _engine;

        private Task _threadActionManager;
        private ActionManager _actionManager;

        internal WindowLayout(CoreWindow cWindow)
        {
            _coreWindow = cWindow;
            _actionManager = new ActionManager(_coreWindow);
            _engine = new DrawEngine(_coreWindow);
            _coreWindow.EventClose += Close;
        }

        private WContainer _window;
        internal WContainer GetContainer()
        {
            return _window;
        }
        internal void SetWindow(WContainer window)
        {
            _window = window;
        }

        //methods
        internal void Show()
        {
            if (!WindowManager.IsRunning())
                WindowManager.StartWith(_coreWindow);
            else
                WindowManager.AddWindow(_coreWindow);
        }

        internal void Close()
        {
            WindowManager.CloseWindow(_coreWindow);
        }

        internal void SetFocusable(bool value)
        {
            _engine.GLWHandler.Focusable = value;
        }

        internal void Minimize()
        {
            _engine.MinimizeRequest = true;
        }

        internal void Maximize()
        {
            _engine.MaximizeRequest = true;
        }

        internal void ToggleFullScreen()
        {
            _engine.FullScreenRequest = true;
        }

        internal void UpdatePosition()
        {
            _engine.UpdatePositionRequest = true;
        }

        internal void UpdateSize()
        {
            _engine.UpdateSizeRequest = true;
        }

        internal void IsFixed(bool flag)
        {
            _window._is_fixed = flag;
        }

        internal void SetEventTask(EventTask task)
        {
            if (!_hold)
                _actionManager.AddTask(task);
        }

        private volatile bool _set = true;

        internal void ExecutePollActions()
        {
            if (!_hold)
            {
                _set = _actionManager.Execute.IsSet;
                if (!_set)
                    _actionManager.Execute.Set();
            }
        }

        internal Prototype GetFocusedItem()
        {
            return _engine.GetFocusedItem();
        }

        internal void SetFocusedItem(Prototype item)
        {
            _engine.SetFocusedItem(item);
        }

        internal void ResetItems()
        {
            _engine.ResetItems();
        }
        internal void ResetFocus()
        {
            _engine.ResetFocus();
        }

        internal void SetIcon(Bitmap icon_big, Bitmap icon_small)
        {
            _engine.SetIcons(icon_big, icon_small);
        }

        internal void SetHidden(bool value)
        {
            _engine.GLWHandler.SetHidden(value);
            _coreWindow.IsHidden = value;
        }

        internal Int64 GetGLWID()
        {
            return _engine.GLWHandler.GetWindowId();
        }

        internal bool IsGLWIDValid()
        {
            if (GetGLWID() == 0)
                return false;
            return true;
        }

        internal bool InitEngine()
        {
            if (_coreWindow.IsHidden)
                SetHidden(false);
            _coreWindow.IsClosed = false;

            _engine.GLWHandler.Transparent = _coreWindow.IsTransparent;
            _engine.GLWHandler.Maximized = _coreWindow.IsMaximized;
            _engine.GLWHandler.Visible = !_coreWindow.IsHidden;
            _engine.GLWHandler.Resizeble = _coreWindow.IsResizable;
            _engine.GLWHandler.BorderHidden = _coreWindow.IsBorderHidden;
            _engine.GLWHandler.AppearInCenter = _coreWindow.IsCentered;
            _engine.GLWHandler.Focusable = _coreWindow.IsFocusable;
            _engine.GLWHandler.AlwaysOnTop = _coreWindow.IsAlwaysOnTop;

            _engine.GLWHandler.GetPointer().SetX(_coreWindow.GetX());
            _engine.GLWHandler.GetPointer().SetY(_coreWindow.GetY());

            _threadActionManager = new Task(() => _actionManager.StartManager());
            _threadActionManager.Start();

            CreateWindowsPair();

            return _engine.Init();
        }

        void CreateWindowsPair()
        {
            WindowsBox.CreateWindowsPair(_coreWindow);
            if (_coreWindow.IsDialog)
            {
                WindowsBox.GetWindowPair(_coreWindow)?.SetFocusable(false);
            }
        }

        internal void Dispose()
        {
            CoreWindow currentPair = GetPairForCurrentWindow();
            DestroyWindowsPair();
            if (_threadActionManager != null && _threadActionManager.Status == TaskStatus.Running)
            {
                _actionManager.StopManager();
                _actionManager.Execute.Set();
            }
            _engine.FreeOnClose();
            if (currentPair != null && !currentPair.IsClosed)
                currentPair.Focus();
        }

        void DestroyWindowsPair()
        {
            if (_coreWindow.IsDialog)
            {
                CoreWindow pair = WindowsBox.GetWindowPair(_coreWindow);
                if (pair != null)
                    pair.SetFocusable(true);
                WindowsBox.RemoveWindow(_coreWindow);
            }
            WindowsBox.RemoveFromWindowDispatcher(_coreWindow);
        }

        internal CoreWindow GetPairForCurrentWindow()
        {
            return WindowsBox.GetWindowPair(_coreWindow);
        }

        internal void UpdateScene()
        {
            _engine.DrawScene();
        }

        internal void Render()
        {
            _engine.Render();
        }

        internal void SetFocus()
        {
            SetFocusable(true);
            _engine.FocusRequest = true;
        }

        private Color _shadeColor = Color.FromArgb(200, 0, 0, 0);

        internal void SetShadeColor(Color color)
        {
            Monitor.Enter(wndLock);
            try
            {
                _shadeColor = Color.FromArgb(color.A, color.R, color.G, color.B);
            }
            finally
            {
                Monitor.Exit(wndLock);
            }
        }

        internal void SetShadeColor(int r, int g, int b)
        {
            SetShadeColor(GraphicsMathService.ColorTransform(r, g, b));
        }

        internal void SetShadeColor(int r, int g, int b, int a)
        {
            SetShadeColor(GraphicsMathService.ColorTransform(r, g, b, a));
        }

        internal void SetShadeColor(float r, float g, float b)
        {
            SetShadeColor(GraphicsMathService.ColorTransform(r, g, b));
        }

        internal void SetShadeColor(float r, float g, float b, float a)
        {
            SetShadeColor(GraphicsMathService.ColorTransform(r, g, b, a));
        }

        internal Color GetShadeColor()
        {
            Monitor.Enter(wndLock);
            try
            {
                return _shadeColor;
            }
            finally
            {
                Monitor.Exit(wndLock);
            }
        }

        internal void FreeVRAMResource<T>(T resource)
        {
            _engine.FreeVRAMResource(resource);
        }

        internal ToolTipItem GetToolTip()
        {
            return _engine.GetToolTip();
        }

        bool _hold = false;

        internal void Hold()
        {
            _hold = true;
        }

        internal void Proceed()
        {
            _hold = false;
        }
    }
}
